lib/upperkut/strategies/scheduled_queue.rb

```diff
@@ -0,0 +1,158 @@
+require 'upperkut/util'
+require 'upperkut/redis_pool'
+require 'upperkut/strategies/base'
+
+module Upperkut
+  module Strategies
+    # Public: Encapsulates methods required to build a Scheculed Queue
+    # Items are queued, but are only fetched at a specific point in time.
+    class ScheduledQueue < Upperkut::Strategies::Base
+      include Upperkut::Util
+
+      ZPOPBYRANGE = %(
+        local score_from = ARGV[1]
+        local score_to = ARGV[2]
+        local limit = ARGV[3]
+
+        local values = redis.call('zrangebyscore', KEYS[1], score_from, score_to, 'LIMIT', '0', limit)
+
+        if table.getn(values) > 0 then
+          redis.call('zrem', KEYS[1], unpack(values))
+        end
+
+        return values
+      ).freeze
+
+      attr_reader :options
+
+      def initialize(worker, options = {})
+        @options = options
+        initialize_options
+        @redis_pool = setup_redis_pool
+        @worker = worker
+      end
+
+      def push_items(items = [])
+        items = [items] if items.is_a?(Hash)
+        return false if items.empty?
+
+        redis do |conn|
+          items.each do |item|
+            ensure_timestamp_attr(item)
+            conn.zadd(key, item['timestamp'], encode_json_item(item))
+          end
+        end
+
+        true
+      end
+
+      def fetch_items
+        args = {
+          value_from: '-inf'.freeze,
+          value_to: Time.now.to_f.to_s,
+          limit: @batch_size
+        }
+        items = []
+
+        redis do |conn|
+          items = pop_values(conn, args)
+        end
+
+        decode_json_items(items)
+      end
+
+      def clear
+        redis { |conn| conn.del(key) }
+      end
+
+      def metrics
+        {
+          'latency' => latency,
+          'size' => size
+        }
+      end
+
+      def process?
+        buff_size = size('-inf', Time.zone.now.to_i)
```

Comment:
Time.zone is a Rails method, not Ruby. Use Time.now.utc.to_i instead.

Reply:
Let's only save events in UTC.

Reply:
:+1: